### PR TITLE
feat: include metadata for LLM models

### LINKED ISSUE #2639: Feature: Include GenAI metadata in model.predict's output
Hi,

[text-bison's API output schema](https://cloud.google.com/vertex-ai/docs/generative-ai/model-reference/text#response_body) contains a metadata field carrying important information:

``` bash
{
  "predictions": [ ... ],
  "metadata": {
    "tokenMetadata": {
      "input_token_count": {
        "total_tokens": integer,
        "total_billable_characters": integer
      },
      "output_token_count": {
        "total_tokens": integer,
        "total_billable_characters": integer
      }
    }
  }
}
```

Currently in the Python SDK, text generation models output a [TextGenerationResponse](https://cloud.google.com/python/docs/reference/aiplatform/latest/vertexai.preview.language_models.TextGenerationResponse) type. And while the response type contains a `_prediction_response` field, this one is a [aiplatform.models.Prediction](https://github.com/googleapis/python-aiplatform/blob/main/vertexai/language_models/_language_models.py#L619C42-L619C70). In other words, it does not contain the additional `metadata` field.

This metadata field is especially important for production-ready use cases, therefore emphasizing the importance of implementing this feature.

Is this of any interest to incorporate in the roadmap?

Thank you,
cymarechal

### PR DESCRIPTION
**Pull request by cymarechal-collibra**
Hi,

This PR includes the metadata output field from Vertex AI Endpoints in `aiplatform.models.Prediction`.
Fixes #2639.

### CODE REVIEW

**Comment by cymarechal-collibra** [comment]
I have signed the CLA.

File: google/cloud/aiplatform/models.py
Comment (Ark-kun) [inline google/cloud/aiplatform/models.py:1605]: Would `json_format.MessageToDict(prediction_response._pb.metadata)` work? (I'm not sure we need to convert the whole response. Also this might better deal with empty metadata.)

**Comment by Ark-kun** [comment]
Thank you for your contribution. Would you be able to add metadata to unit tests (in `test_models.py`)

**Comment by cymarechal-collibra** [comment]
Hi @Ark-kun,
Thank you for your quick review and suggestions.
I adapted the tests in `test_endpoints.py` as `test_models.py` does not seem to test for model output.
Please let me know if I can help with anything else.

**Review by Ark-kun — APPROVED** [review]

**Comment by Ark-kun** [comment]
The tests seem to be failing:
```
    @pytest.mark.asyncio
    @pytest.mark.usefixtures("get_endpoint_mock")
    async def test_predict_async(self, predict_async_client_predict_mock):
        """Tests the Endpoint.predict_async method."""
        test_endpoint = models.Endpoint(_TEST_ID)
        test_prediction = await test_endpoint.predict_async(
            instances=_TEST_INSTANCES, parameters={"param": 3.0}
        )

        true_prediction = models.Prediction(
            predictions=_TEST_PREDICTION,
            deployed_model_id=_TEST_ID,
            metadata=_TEST_METADATA,
            model_version_id=_TEST_VERSION_ID,
            model_resource_name=_TEST_MODEL_NAME,
        )

>       assert true_prediction == test_prediction
E       AssertionError: assert Prediction(pr...anations=None) == Prediction(pr...anations=None)
E
E         Omitting 5 identical items, use -vv to show
E         Differing attributes:
E         ['metadata']
E
E         Drill down into differing attribute metadata:
E           metadata: {'foo': 'bar'} != None
E         Use -v to get more diff
```

I think you need to change the `predict_async` method. And maybe the `explain` and `explain_async` as well.

And we should be good to go.

**Comment by cymarechal-collibra** [comment]
Urg, sorry about that. The tests were silently skipped on my end.

- I edited the relevant tests;
- I included the metadata field on the response of the raw predict method;
- However, I did not include it on the explain method(s) as [the metadata field is not included in the protobuf definition](https://github.com/googleapis/python-aiplatform/blob/f9782007c58ee11fe276c373d8d7ac6c2b0cb249/google/cloud/aiplatform_v1/types/prediction_service.py#L319).

**Review by Ark-kun — APPROVED** [review]

**Comment by Ark-kun** [comment]
@cymarechal-collibra I apologize for asking you to do one more thing due to the current process. This repo has a formatting presubmit (using `black`) and it needs to pass for the change to be pulled in. Can you please make this small formatting change?
```
nox > black --check --diff docs google tests noxfile.py setup.py
--- tests/unit/aiplatform/test_endpoints.py	2023-10-24 19:43:18 +0000
+++ tests/unit/aiplatform/test_endpoints.py	2023-10-24 20:05:00.424203 +0000
@@ -606,14 +606,16 @@
 @pytest.fixture
 def predict_private_endpoint_mock():
     with mock.patch.object(urllib3.PoolManager, "request") as predict_mock:
         predict_mock.return_value = urllib3.response.HTTPResponse(
             status=200,
-            body=json.dumps({
-                "predictions": _TEST_PREDICTION,
-                "metadata": _TEST_METADATA,
-            }),
+            body=json.dumps(
+                {
+                    "predictions": _TEST_PREDICTION,
+                    "metadata": _TEST_METADATA,
+                }
+            ),
         )
         yield predict_mock


 @pytest.fixture
would reformat tests/unit/aiplatform/test_endpoints.py

Oh no! 💥 💔 💥
1 file would be reformatted, 975 files would be left unchanged.
```

**Comment by meken** [comment]
Is this still waiting for the linting issue to be fixed before it can be merged? It looks like this PR hasn't been updated for more than 2 months, while the functionality is very much needed. If @cymarechal-collibra can't complete it, I'm open to apply his changes in another fork.

**Comment by matthew29tang** [comment]
Due to inactivity, I've merged in #3301 for this feature request.

**Pull request closed without merging**
